In [1]:
import numpy as np
import pandas as pd
import geopandas as gpd

from scipy.sparse import hstack, csr_matrix
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import NearestNeighbors
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_absolute_percentage_error, root_mean_squared_error

RANDOM_STATE = 42
TARGET = "ClosePrice"

## 0. Load Cleaned CRMLS and California School District Datasets

In [2]:
training_set = pd.read_parquet("../data/train_preprocessed.parquet")
testing_set = pd.read_parquet("../data/test_preprocessed.parquet")
train_df, test_df = training_set.copy(), testing_set.copy()
df = pd.read_parquet("../data/full_data_preprocessed.parquet")
original_models = pd.read_parquet("../data/baseline_model_performances.parquet")
print(f"Train dataset: {train_df.shape}, Test dataset: {test_df.shape}, Full Dataset: {df.shape} \n")

districts_raw = gpd.read_file("../data/California_School_District_Areas_2024-25.geojson")
print(f"Loaded {len(districts_raw)} districts from California School District Areas 2024-2025")
print(f"District Type Unique Entries: {districts_raw["DistrictType"].unique()}")

Train dataset: (81857, 30), Test dataset: (7442, 30), Full Dataset: (240901, 30) 

Loaded 937 districts from California School District Areas 2024-2025
District Type Unique Entries: ['Unified' 'Elementary' 'High']


In [3]:
districts = districts_raw[districts_raw["DistrictType"].isin(["Unified", "High"])].copy()
districts = districts[["DistrictType", "DistrictName", "geometry"]].to_crs("EPSG:4326")
print(f"Filtered to {len(districts)} Unified/High polygons")

Filtered to 421 Unified/High polygons


In [4]:
# raw unscaled lookup table
raw_lookup = df[["ListingKey", "BedroomsTotal", "BathroomsTotalInteger", "YearBuilt", "Latitude", "Longitude"]]
raw_lookup = raw_lookup.rename(columns={
    "BedroomsTotal": "BedroomsTotal_raw",
    "BathroomsTotalInteger": "BathroomsTotalInteger_raw",
    "YearBuilt": "YearBuilt_raw",
    "Latitude": "Latitude_raw",
    "Longitude": "Longitude_raw",
})
n_before = len(raw_lookup)
raw_lookup = raw_lookup.drop_duplicates(subset="ListingKey", keep="first")
print(f"raw_lookup: {len(raw_lookup)} rows ({n_before - len(raw_lookup)} duplicate ListingKeys removed)")

raw_lookup: 240699 rows (202 duplicate ListingKeys removed)


#### Helper Functions:

In [5]:
def merge_raw_lookup(frame, lookup=raw_lookup):
    n_before = len(frame)
    merged = frame.merge(lookup, on="ListingKey", how="left")
    assert len(merged) == n_before, f"Merge changed row count: {n_before} -> {len(merged)} (duplicate key in lookup)"
    merged.index = frame.index
    return merged

# Build X and y train and test sets
encoder = OneHotEncoder(handle_unknown="ignore", sparse_output=True)

def build_X_y(frame, NON_FEATURE_COLS, CATEGORICAL_COLS):
    encoder.fit(train_df[CATEGORICAL_COLS])
    numeric_part = frame.drop(columns=[c for c in [TARGET] + NON_FEATURE_COLS + CATEGORICAL_COLS if c in frame.columns])
    numeric_part = numeric_part.select_dtypes(include=[np.number])
    encoded = encoder.transform(frame[CATEGORICAL_COLS])
    X = hstack([csr_matrix(numeric_part.values), encoded]).tocsr()
    y = frame[TARGET].values
    feature_names = list(numeric_part.columns) + list(encoder.get_feature_names_out(CATEGORICAL_COLS))
    return X, y, feature_names

# Evaluate Model Performance Function
def evaluate_models(X_train,y_train,X_test,y_test, NEW_FEATURE):
    def evaluate_model(model, X_train, y_train, X_test, y_test, name):
        train_pred, test_pred = model.predict(X_train), model.predict(X_test)
        return {"model": name, "train_r2": r2_score(y_train, train_pred), "test_r2": r2_score(y_test, test_pred),
            "test_mae": mean_absolute_error(y_test, test_pred), "test_rmse": root_mean_squared_error(y_test, test_pred),
            "test_mape": mean_absolute_percentage_error(y_test, test_pred)}

    new_results = []
    linear_model = LinearRegression().fit(X_train, y_train)
    new_results.append(evaluate_model(linear_model, X_train, y_train, X_test, y_test, "linear regression"))

    tree_tuned = DecisionTreeRegressor(max_depth=12, min_samples_leaf=20, random_state=RANDOM_STATE).fit(X_train, y_train)
    new_results.append(evaluate_model(tree_tuned, X_train, y_train, X_test, y_test, "decision tree"))

    rf_model = RandomForestRegressor(n_estimators=200, max_depth=20, min_samples_leaf=5, n_jobs=-1, random_state=RANDOM_STATE).fit(X_train, y_train)
    new_results.append(evaluate_model(rf_model, X_train, y_train, X_test, y_test, "random forest"))

    new_results_df = pd.DataFrame(new_results)
    new_results_df["New Features"] = [NEW_FEATURE for i in range(len(new_results_df))]

    return new_results_df.round(4)

## 1. Add Geographic School District Layer

In [6]:
def add_spatial_district(frame, districts):
    points_gdf = gpd.GeoDataFrame(
        geometry=gpd.points_from_xy(frame["Longitude_raw"], frame["Latitude_raw"]),
        crs="EPSG:4326",
        index=frame.index,
    )
    joined = gpd.sjoin(points_gdf, districts, how="left", predicate="within")
    joined = joined[~joined.index.duplicated(keep="first")]

    spatial_col = pd.Series(np.nan, index=frame.index, dtype=object)
    spatial_col.loc[joined.index] = joined["DistrictName"].values
    print(f"Real spatial match rate (before fallback): {spatial_col.notna().mean():.2%}")
    return spatial_col.fillna(frame["HighSchoolDistrict"])

train_df = merge_raw_lookup(train_df)
test_df = merge_raw_lookup(test_df)
train_df["SchoolDistrictSpatial"] = add_spatial_district(train_df,districts) # unedited Training and Testing Sets
test_df["SchoolDistrictSpatial"] = add_spatial_district(test_df, districts)

print(f"Train coverage: {train_df['SchoolDistrictSpatial'].notna().mean():.2%}")
print(f"Test coverage:  {test_df['SchoolDistrictSpatial'].notna().mean():.2%}")

Real spatial match rate (before fallback): 99.89%
Real spatial match rate (before fallback): 99.95%
Train coverage: 100.00%
Test coverage:  100.00%


In [7]:
NON_FEATURE_COLS = ["ListingKey", "CloseDate", "BedroomsTotal_raw", "BathroomsTotalInteger_raw", "YearBuilt_raw"]
CATEGORICAL_COLS = ["City", "PostalCode", "MLSAreaMajor", "HighSchoolDistrict", "SchoolDistrictSpatial", "Levels"]
CATEGORICAL_COLS = [c for c in CATEGORICAL_COLS if c in train_df.columns]

X_train, y_train, feature_names = build_X_y(train_df,NON_FEATURE_COLS,CATEGORICAL_COLS)
X_test, y_test, _ = build_X_y(test_df,NON_FEATURE_COLS,CATEGORICAL_COLS)

# Create Comparison df to track model performances as features are added
og_models = original_models.copy()
og_models["New Features"] = ["Baseline" for i in range(len(og_models))]

comparison_df = pd.concat([og_models,evaluate_models(X_train, y_train, X_test, y_test, "SchoolDistrictSpatial")], ignore_index = True)
comparison_df

,model,train_r2,test_r2,test_mae,test_rmse,test_mape,New Features
0,linear regression,0.8534,0.8420,187946.0976,313321.9877,0.1904,Baseline
1,decision tree,0.8418,0.7961,195030.9801,355897.5189,0.1663,Baseline
2,random forest,0.9366,0.8734,143843.9497,280387.4237,0.1190,Baseline
3,linear regression,0.8547,0.8428,187109.9833,312499.9337,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,195612.6055,356542.6947,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,143588.6477,280461.0095,0.1188,SchoolDistrictSpatial


In [8]:
agreement = (train_df["SchoolDistrictSpatial"] == train_df["HighSchoolDistrict"]).mean()
print(f"SchoolDistrictSpatial matches self-reported HighSchoolDistrict: {agreement:.1%}")

n_spatial = train_df["SchoolDistrictSpatial"].nunique()
n_self_reported = train_df["HighSchoolDistrict"].nunique()
print(f"Distinct spatial districts: {n_spatial}, distinct self-reported: {n_self_reported}")

mismatched = train_df[train_df["SchoolDistrictSpatial"] != train_df["HighSchoolDistrict"]]
mismatched[["HighSchoolDistrict", "SchoolDistrictSpatial"]].value_counts().head(20)

SchoolDistrictSpatial matches self-reported HighSchoolDistrict: 58.2%
Distinct spatial districts: 377, distinct self-reported: 405


HighSchoolDistrict        SchoolDistrictSpatial         
Unknown                   Desert Sands Unified              1652
William S. Hart Union     William S. Hart Union High        1496
Unknown                   San Diego Unified                 1428
Temecula Unified          Temecula Valley Unified           1187
Unknown                   Palm Springs Unified              1116
Antelope Valley Union     Antelope Valley Union High        1100
Menifee Union             Perris Union High                  871
Murrieta                  Murrieta Valley Unified            829
Newport Mesa Unified      Newport-Mesa Unified               787
Beaumont                  Beaumont Unified                   732
Unknown                   Los Angeles Unified                726
                          Grossmont Union High               705
Victor Valley Unified     Victor Valley Union High           663
Unknown                   Poway Unified                      488
                          Coachel

- Spatial feature is important when HighSchoolDistrict. is "Unknown"
- redundant when the difference is only in naming convention
- And very few convert High School and Elementary districts

In [9]:
def normalize_district_name(series):
    return (series.str.lower()
            .str.replace(r"\b(unified|union|high|school|district|joint)\b", "", regex=True)
            .str.replace(r"[^a-z]", "", regex=True))

has_self_reported = train_df["HighSchoolDistrict"].notna() & (train_df["HighSchoolDistrict"] != "Unknown")

filled_missing = (~has_self_reported).sum()
print(f"Rows where spatial join filled in a previously missing/Unknown value: {filled_missing} ({filled_missing/len(train_df):.1%})")

both_present = train_df[has_self_reported]
normalized_match_both_present = (
    normalize_district_name(both_present["HighSchoolDistrict"]) ==
    normalize_district_name(both_present["SchoolDistrictSpatial"])
)
print(f"True agreement rate, self-reported vs spatial, when both have an answer: {normalized_match_both_present.mean():.1%}")

Rows where spatial join filled in a previously missing/Unknown value: 11221 (13.7%)
True agreement rate, self-reported vs spatial, when both have an answer: 82.9%


In [10]:
CATEGORICAL_COLS = ["City", "PostalCode", "MLSAreaMajor", "SchoolDistrictSpatial", "Levels"]  # HighSchoolDistrict replaced, not added alongside
CATEGORICAL_COLS = [c for c in CATEGORICAL_COLS if c in train_df.columns]

X_train, y_train, feature_names = build_X_y(train_df, NON_FEATURE_COLS, CATEGORICAL_COLS)
X_test, y_test, _ = build_X_y(test_df, NON_FEATURE_COLS, CATEGORICAL_COLS)

comparison_df = pd.concat([comparison_df, evaluate_models(
    X_train, y_train, X_test, y_test, "SchoolDistrictSpatial (replacing HighSchoolDistrict)"
)], ignore_index=True)
comparison_df

,model,train_r2,test_r2,test_mae,test_rmse,test_mape,New Features
0,linear regression,0.8534,0.8420,187946.0976,313321.9877,0.1904,Baseline
1,decision tree,0.8418,0.7961,195030.9801,355897.5189,0.1663,Baseline
2,random forest,0.9366,0.8734,143843.9497,280387.4237,0.1190,Baseline
3,linear regression,0.8547,0.8428,187109.9833,312499.9337,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,195612.6055,356542.6947,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,143588.6477,280461.0095,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,186061.2210,311142.1399,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,195510.3116,357268.4797,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,143345.3735,280532.5473,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...


#### Imrpovement for the linear model with no meaningful cost elsewhere when joining HighSchoolDistrict (MLS origin) with SchoolDistrictSpatial (CA School District Data)
- Linear Regression test R2 went from 0.8420 -> 0.8442
- Decision Tree dipped slightly (0.7961 -> 0.7945, likely noise), 
- Random Forest is essentially flat with a small MAPE improvement. 

## 2. Create Bed to Bath Ratio Feature

In [11]:
# reset train_df and test_df after engineering the feature above
train_df, test_df = training_set.copy(), testing_set.copy()
train_df = merge_raw_lookup(train_df)
test_df = merge_raw_lookup(test_df)

In [12]:
# Engineer Bed to Bath Ratio Feature
train_df["BedBathRatio"] = train_df["BedroomsTotal_raw"] / train_df["BathroomsTotalInteger_raw"].replace(0, np.nan)
test_df["BedBathRatio"] = test_df["BedroomsTotal_raw"] / test_df["BathroomsTotalInteger_raw"].replace(0, np.nan)

train_bed_bath_median = train_df["BedBathRatio"].median()
train_df["BedBathRatio"] = train_df["BedBathRatio"].fillna(train_bed_bath_median)
test_df["BedBathRatio"] = test_df["BedBathRatio"].fillna(train_bed_bath_median)

In [13]:
CATEGORICAL_COLS = ["City", "PostalCode", "MLSAreaMajor", "HighSchoolDistrict", "Levels"] # Exclude Geographic layer engineered above
CATEGORICAL_COLS = [c for c in CATEGORICAL_COLS if c in train_df.columns] 
# Tests only Bed Bath Ratio

X_train, y_train, feature_names = build_X_y(train_df,NON_FEATURE_COLS,CATEGORICAL_COLS)
X_test, y_test, _ = build_X_y(test_df,NON_FEATURE_COLS,CATEGORICAL_COLS)

comparison_df = pd.concat([comparison_df,evaluate_models(X_train, y_train, X_test, y_test, "BedBathRatio")], ignore_index = True )
comparison_df

,model,train_r2,test_r2,test_mae,test_rmse,test_mape,New Features
0,linear regression,0.8534,0.8420,187946.0976,313321.9877,0.1904,Baseline
1,decision tree,0.8418,0.7961,195030.9801,355897.5189,0.1663,Baseline
2,random forest,0.9366,0.8734,143843.9497,280387.4237,0.1190,Baseline
3,linear regression,0.8547,0.8428,187109.9833,312499.9337,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,195612.6055,356542.6947,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,143588.6477,280461.0095,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,186061.2210,311142.1399,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,195510.3116,357268.4797,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,143345.3735,280532.5473,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...
9,linear regression,0.8576,0.8459,184492.5745,309355.5185,0.1835,BedBathRatio


#### Improvement for Linear Model, and Decision Tree and Random Forest Models stayed Roughly the same

## 3. Create Property Age feature

In [14]:
# reset train_df and test_df after engineering the features above
train_df, test_df = training_set.copy(), testing_set.copy()
train_df = merge_raw_lookup(train_df)
test_df = merge_raw_lookup(test_df)

In [15]:
# Engineer Propert Age Feature
train_df["PropertyAgeAtSale"] = (train_df["CloseDate"].dt.year - train_df["YearBuilt_raw"]).clip(lower=0)
test_df["PropertyAgeAtSale"] = (test_df["CloseDate"].dt.year - test_df["YearBuilt_raw"]).clip(lower=0)

train_property_age_median = train_df["PropertyAgeAtSale"].median()
train_df["PropertyAgeAtSale"] = train_df["PropertyAgeAtSale"].fillna(train_property_age_median)
test_df["PropertyAgeAtSale"] = test_df["PropertyAgeAtSale"].fillna(train_property_age_median)

In [16]:
CATEGORICAL_COLS = ["City", "PostalCode", "MLSAreaMajor", "HighSchoolDistrict", "Levels"] 
CATEGORICAL_COLS = [c for c in CATEGORICAL_COLS if c in train_df.columns] 
# Tests only Property Age At Sale

X_train, y_train, feature_names = build_X_y(train_df,NON_FEATURE_COLS,CATEGORICAL_COLS)
X_test, y_test, _ = build_X_y(test_df,NON_FEATURE_COLS,CATEGORICAL_COLS)

comparison_df = pd.concat([comparison_df,evaluate_models(X_train, y_train, X_test, y_test, "PropertyAgeAtSale")], ignore_index = True )
comparison_df

,model,train_r2,test_r2,test_mae,test_rmse,test_mape,New Features
0,linear regression,0.8534,0.8420,187946.0976,313321.9877,0.1904,Baseline
1,decision tree,0.8418,0.7961,195030.9801,355897.5189,0.1663,Baseline
2,random forest,0.9366,0.8734,143843.9497,280387.4237,0.1190,Baseline
3,linear regression,0.8547,0.8428,187109.9833,312499.9337,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,195612.6055,356542.6947,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,143588.6477,280461.0095,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,186061.2210,311142.1399,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,195510.3116,357268.4797,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,143345.3735,280532.5473,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...
9,linear regression,0.8576,0.8459,184492.5745,309355.5185,0.1835,BedBathRatio


#### Slight Improvement for Linear Regression and Decision Tree, negligble change in Random Forest

## 4. K-Nearest Neighbors Comparable Price (CompPriceKNN)

In [17]:
# reset train_df and test_df after engineering the features above
train_df, test_df = training_set.copy(), testing_set.copy()
train_df = merge_raw_lookup(train_df)
test_df = merge_raw_lookup(test_df)

COMP_K = 15     # nearest 15 neighbors by long and lat
coord_cols = ["Latitude_raw", "Longitude_raw"]
train_coords = train_df[coord_cols].values
train_prices = train_df["ClosePrice"].values

nn = NearestNeighbors(n_neighbors=COMP_K + 1, algorithm="ball_tree").fit(train_coords)
_, neighbor_idx = nn.kneighbors(train_coords)
neighbor_idx = neighbor_idx[:, 1:]  # drop self
train_df["CompPriceKNN"] = train_prices[neighbor_idx].mean(axis=1)

nn_test = NearestNeighbors(n_neighbors=COMP_K, algorithm="ball_tree").fit(train_coords)
_, test_neighbor_idx = nn_test.kneighbors(test_df[coord_cols].values)
test_df["CompPriceKNN"] = train_prices[test_neighbor_idx].mean(axis=1)

In [18]:
# CompPriceKNN would greatly improve DT and RF, but heavily reduce LR because 
# it had massive gaps in value between columns, unlike other previously normalized numeric features

comp_scaler = StandardScaler()
train_df["CompPriceKNN"] = comp_scaler.fit_transform(train_df[["CompPriceKNN"]])
test_df["CompPriceKNN"] = comp_scaler.transform(test_df[["CompPriceKNN"]])

In [19]:
NON_FEATURE_COLS = ["ListingKey", "CloseDate", "BedroomsTotal_raw", "BathroomsTotalInteger_raw", "YearBuilt_raw", "Latitude_raw", "Longitude_raw"]
CATEGORICAL_COLS = ["City", "PostalCode", "MLSAreaMajor", "HighSchoolDistrict", "Levels"] 
CATEGORICAL_COLS = [c for c in CATEGORICAL_COLS if c in train_df.columns] 
# Tests only Property Age 

X_train, y_train, feature_names = build_X_y(train_df,NON_FEATURE_COLS,CATEGORICAL_COLS)
X_test, y_test, _ = build_X_y(test_df,NON_FEATURE_COLS,CATEGORICAL_COLS)

comparison_df = pd.concat([comparison_df,evaluate_models(X_train, y_train, X_test, y_test, "CompPriceKNN")])
comparison_df

,model,train_r2,test_r2,test_mae,test_rmse,test_mape,New Features
0,linear regression,0.8534,0.8420,187946.0976,313321.9877,0.1904,Baseline
1,decision tree,0.8418,0.7961,195030.9801,355897.5189,0.1663,Baseline
2,random forest,0.9366,0.8734,143843.9497,280387.4237,0.1190,Baseline
3,linear regression,0.8547,0.8428,187109.9833,312499.9337,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,195612.6055,356542.6947,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,143588.6477,280461.0095,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,186061.2210,311142.1399,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,195510.3116,357268.4797,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,143345.3735,280532.5473,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...
9,linear regression,0.8576,0.8459,184492.5745,309355.5185,0.1835,BedBathRatio


#### Imrpovements for all models (greatest change in DT, then LR, then RF test R2)
R2 Change:
- LR: 0.8420 -> 0.8589
- DT: 0.7961 -> 0.8552
- RF: 0.8734 -> 0.8850

## 5. Add MonthSin and MonthCos (Seasonality) and MonthsSinceStart (Linear Trend)

In [20]:
# Using Fourier Terms:
# Cyclical time encoding lets the model recognize December and January as adjacent rather than maximally distant (like 1 - 12 would be).

# reset train_df and test_df after engineering the features above
train_df, test_df = training_set.copy(), testing_set.copy()
train_df = merge_raw_lookup(train_df)
test_df = merge_raw_lookup(test_df)

close_month_period = pd.PeriodIndex(train_df["CloseMonth"], freq="M")
test_month_period = pd.PeriodIndex(test_df["CloseMonth"], freq="M")

all_months_sorted = sorted(set(close_month_period) | set(test_month_period))
month_to_trend = {m: i for i, m in enumerate(all_months_sorted)}

train_df["MonthsSinceStart"] = close_month_period.map(month_to_trend)
test_df["MonthsSinceStart"] = test_month_period.map(month_to_trend)

train_df["MonthSin"] = np.sin(2 * np.pi * close_month_period.month / 12)
train_df["MonthCos"] = np.cos(2 * np.pi * close_month_period.month / 12)
test_df["MonthSin"] = np.sin(2 * np.pi * test_month_period.month / 12)
test_df["MonthCos"] = np.cos(2 * np.pi * test_month_period.month / 12)

In [21]:
CATEGORICAL_COLS = ["City", "PostalCode", "MLSAreaMajor", "HighSchoolDistrict", "Levels"] 
CATEGORICAL_COLS = [c for c in CATEGORICAL_COLS if c in train_df.columns] 

X_train, y_train, feature_names = build_X_y(train_df,NON_FEATURE_COLS,CATEGORICAL_COLS)
X_test, y_test, _ = build_X_y(test_df,NON_FEATURE_COLS,CATEGORICAL_COLS)

comparison_df = pd.concat([comparison_df,evaluate_models(X_train, y_train, X_test, y_test, "Cyclical Time Encoding")])
comparison_df

,model,train_r2,test_r2,test_mae,test_rmse,test_mape,New Features
0,linear regression,0.8534,0.8420,187946.0976,313321.9877,0.1904,Baseline
1,decision tree,0.8418,0.7961,195030.9801,355897.5189,0.1663,Baseline
2,random forest,0.9366,0.8734,143843.9497,280387.4237,0.1190,Baseline
3,linear regression,0.8547,0.8428,187109.9833,312499.9337,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,195612.6055,356542.6947,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,143588.6477,280461.0095,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,186061.2210,311142.1399,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,195510.3116,357268.4797,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,143345.3735,280532.5473,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...
9,linear regression,0.8576,0.8459,184492.5745,309355.5185,0.1835,BedBathRatio


#### With only one year of training data, there's no repetition for the model to learn a seasonal pattern from
This explains the slight changes in the model performances. We can experiment with this feature's implementation and the total data we use for training for more significant results.

In [22]:
# comparison df cleaning 
comparison_df.drop(columns=['test_mae','test_rmse'])

,model,train_r2,test_r2,test_mape,New Features
0,linear regression,0.8534,0.8420,0.1904,Baseline
1,decision tree,0.8418,0.7961,0.1663,Baseline
2,random forest,0.9366,0.8734,0.1190,Baseline
3,linear regression,0.8547,0.8428,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...
9,linear regression,0.8576,0.8459,0.1835,BedBathRatio


| Model | Train R^2 | Test R^2 | Test MAPE | New Features | Test R^2 Difference |
| :--- | :--- | :--- | :--- | :--- | :--- |
| linear regression | 0.8534 | 0.8420 | 0.1904 | Baseline | 0.0000 |
| decision tree | 0.8418 | 0.7961 | 0.1663 | Baseline | 0.0000 |
| random forest | 0.9366 | 0.8734 | 0.1190 | Baseline | 0.0000 |
| linear regression | 0.8535 | 0.8418 | 0.1905 | SchoolDistrictSpatial | -0.0002 |
| decision tree | 0.8418 | 0.7961 | 0.1663 | SchoolDistrictSpatial | 0.0000 |
| random forest | 0.9366 | 0.8736 | 0.1190 | SchoolDistrictSpatial | +0.0002 |
| linear regression | 0.8577 | 0.8459 | 0.1835 | BedBathRatio | +0.0039 |
| decision tree | 0.8422 | 0.7961 | 0.1662 | BedBathRatio | 0.0000 |
| random forest | 0.9377 | 0.8739 | 0.1186 | BedBathRatio | +0.0005 |
| linear regression | 0.8536 | 0.8428 | 0.1890 | PropertyAgeAtSale | +0.0008 |
| decision tree | 0.8417 | 0.7971 | 0.1660 | PropertyAgeAtSale | +0.0010 |
| random forest | 0.9374 | 0.8740 | 0.1188 | PropertyAgeAtSale | +0.0006 |
| linear regression | 0.8069 | 0.8093 | 0.1961 | CompPriceKNN | -0.0327 |
| decision tree | 0.8923 | 0.8536 | 0.1338 | CompPriceKNN | +0.0575 |
| random forest | 0.9471 | 0.8852 | 0.1129 | CompPriceKNN | +0.0118 |
| linear regression | 0.8537 | 0.8424 | 0.1889 | Cyclical Time Encoding | +0.0004 |
| decision tree | 0.8419 | 0.7967 | 0.1666 | Cyclical Time Encoding | +0.0006 |
| random forest | 0.9376 | 0.8726 | 0.1202 | Cyclical Time Encoding | -0.0008 |

#### Linear Regression and Decision Tree models barely improved, while Random Forest model performance slighty declined

# 6. Adding all features

In [25]:
# reset DataFrames
train_df, test_df = training_set.copy(), testing_set.copy()
train_df = merge_raw_lookup(train_df)
test_df = merge_raw_lookup(test_df)

# raw unscaled lookup table
raw_lookup = df[["ListingKey", "BedroomsTotal", "BathroomsTotalInteger", "YearBuilt", "Latitude", "Longitude"]]
raw_lookup = raw_lookup.rename(columns={
    "BedroomsTotal": "BedroomsTotal_raw",
    "BathroomsTotalInteger": "BathroomsTotalInteger_raw",
    "YearBuilt": "YearBuilt_raw",
    "Latitude": "Latitude_raw",
    "Longitude": "Longitude_raw",
})
n_before = len(raw_lookup)
raw_lookup = raw_lookup.drop_duplicates(subset="ListingKey", keep="first")
print(f"raw_lookup: {len(raw_lookup)} rows ({n_before - len(raw_lookup)} duplicate ListingKeys removed)")

raw_lookup: 240699 rows (202 duplicate ListingKeys removed)


In [26]:
def add_spatial_district(frame, districts):
    points_gdf = gpd.GeoDataFrame(
        geometry=gpd.points_from_xy(frame["Longitude_raw"], frame["Latitude_raw"]),
        crs="EPSG:4326",
        index=frame.index,
    )
    joined = gpd.sjoin(points_gdf, districts, how="left", predicate="within")
    joined = joined[~joined.index.duplicated(keep="first")]

    spatial_col = pd.Series(np.nan, index=frame.index, dtype=object)
    spatial_col.loc[joined.index] = joined["DistrictName"].values
    print(f"Real spatial match rate (before fallback): {spatial_col.notna().mean():.2%}")
    return spatial_col.fillna(frame["HighSchoolDistrict"])

train_df["SchoolDistrictSpatial"] = add_spatial_district(train_df,districts) # unedited Training and Testing Sets
test_df["SchoolDistrictSpatial"] = add_spatial_district(test_df, districts)

Real spatial match rate (before fallback): 99.89%
Real spatial match rate (before fallback): 99.95%


In [27]:
train_df["BedBathRatio"] = train_df["BedroomsTotal_raw"] / train_df["BathroomsTotalInteger_raw"].replace(0, np.nan)
test_df["BedBathRatio"] = test_df["BedroomsTotal_raw"] / test_df["BathroomsTotalInteger_raw"].replace(0, np.nan)

train_bed_bath_median = train_df["BedBathRatio"].median()
train_df["BedBathRatio"] = train_df["BedBathRatio"].fillna(train_bed_bath_median)
test_df["BedBathRatio"] = test_df["BedBathRatio"].fillna(train_bed_bath_median)

train_df["PropertyAgeAtSale"] = (train_df["CloseDate"].dt.year - train_df["YearBuilt_raw"]).clip(lower=0)
test_df["PropertyAgeAtSale"] = (test_df["CloseDate"].dt.year - test_df["YearBuilt_raw"]).clip(lower=0)

train_property_age_median = train_df["PropertyAgeAtSale"].median()
train_df["PropertyAgeAtSale"] = train_df["PropertyAgeAtSale"].fillna(train_property_age_median)
test_df["PropertyAgeAtSale"] = test_df["PropertyAgeAtSale"].fillna(train_property_age_median)

In [28]:
COMP_K = 15     # nearest 15 neighbors by long and lat
coord_cols = ["Latitude_raw", "Longitude_raw"]
train_coords = train_df[coord_cols].values
train_prices = train_df["ClosePrice"].values

nn = NearestNeighbors(n_neighbors=COMP_K + 1, algorithm="ball_tree").fit(train_coords)
_, neighbor_idx = nn.kneighbors(train_coords)
neighbor_idx = neighbor_idx[:, 1:]  # drop self
train_df["CompPriceKNN"] = train_prices[neighbor_idx].mean(axis=1)

nn_test = NearestNeighbors(n_neighbors=COMP_K, algorithm="ball_tree").fit(train_coords)
_, test_neighbor_idx = nn_test.kneighbors(test_df[coord_cols].values)
test_df["CompPriceKNN"] = train_prices[test_neighbor_idx].mean(axis=1)

comp_scaler = StandardScaler()
train_df["CompPriceKNN"] = comp_scaler.fit_transform(train_df[["CompPriceKNN"]])
test_df["CompPriceKNN"] = comp_scaler.transform(test_df[["CompPriceKNN"]])

In [29]:
close_month_period = pd.PeriodIndex(train_df["CloseMonth"], freq="M")
test_month_period = pd.PeriodIndex(test_df["CloseMonth"], freq="M")

all_months_sorted = sorted(set(close_month_period) | set(test_month_period))
month_to_trend = {m: i for i, m in enumerate(all_months_sorted)}

train_df["MonthsSinceStart"] = close_month_period.map(month_to_trend)
test_df["MonthsSinceStart"] = test_month_period.map(month_to_trend)

train_df["MonthSin"] = np.sin(2 * np.pi * close_month_period.month / 12)
train_df["MonthCos"] = np.cos(2 * np.pi * close_month_period.month / 12)
test_df["MonthSin"] = np.sin(2 * np.pi * test_month_period.month / 12)
test_df["MonthCos"] = np.cos(2 * np.pi * test_month_period.month / 12)

In [ ]:
NON_FEATURE_COLS = ["ListingKey", "CloseDate", "BedroomsTotal_raw", "BathroomsTotalInteger_raw", "YearBuilt_raw", "Latitude_raw", "Longitude_raw"]
CATEGORICAL_COLS = ["City", "PostalCode", "MLSAreaMajor", "SpatialSchoolDistrict"] 
CATEGORICAL_COLS = [c for c in CATEGORICAL_COLS if c in train_df.columns] 

X_train, y_train, feature_names = build_X_y(train_df,NON_FEATURE_COLS,CATEGORICAL_COLS)
X_test, y_test, _ = build_X_y(test_df,NON_FEATURE_COLS,CATEGORICAL_COLS)

comparison_df = pd.concat([comparison_df,evaluate_models(X_train, y_train, X_test, y_test, "Engineered Features")])

,model,train_r2,test_r2,test_mae,test_rmse,test_mape,New Features
0,linear regression,0.8534,0.8420,187946.0976,313321.9877,0.1904,Baseline
1,decision tree,0.8418,0.7961,195030.9801,355897.5189,0.1663,Baseline
2,random forest,0.9366,0.8734,143843.9497,280387.4237,0.1190,Baseline
3,linear regression,0.8547,0.8428,187109.9833,312499.9337,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,195612.6055,356542.6947,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,143588.6477,280461.0095,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,186061.2210,311142.1399,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,195510.3116,357268.4797,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,143345.3735,280532.5473,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...
9,linear regression,0.8576,0.8459,184492.5745,309355.5185,0.1835,BedBathRatio


In [ ]:
comparison_df = comparison_df.drop(columns=['test_mae','test_rmse'])

,model,train_r2,test_r2,test_mape,New Features
0,linear regression,0.8534,0.8420,0.1904,Baseline
1,decision tree,0.8418,0.7961,0.1663,Baseline
2,random forest,0.9366,0.8734,0.1190,Baseline
3,linear regression,0.8547,0.8428,0.1891,SchoolDistrictSpatial
4,decision tree,0.8416,0.7954,0.1666,SchoolDistrictSpatial
5,random forest,0.9368,0.8734,0.1188,SchoolDistrictSpatial
6,linear regression,0.8535,0.8442,0.1876,SchoolDistrictSpatial (replacing HighSchoolDis...
7,decision tree,0.8414,0.7945,0.1664,SchoolDistrictSpatial (replacing HighSchoolDis...
8,random forest,0.9366,0.8733,0.1182,SchoolDistrictSpatial (replacing HighSchoolDis...
9,linear regression,0.8576,0.8459,0.1835,BedBathRatio


# 7. Work In Progress

- Wildfire risk zone feature (using same idea from District Geographic Layer)
- Potential flooding zones (same as above if applicable if data is available)
- See if there are ratings, scorings, or potential sentiment analysis data for each school district used in the district geographic layer
- Interaction Terms for the Linear Regression Model
- Change CompPriceKNN to account for neighbors having larger or smaller houses than the listing (by dividing by living area or lotsize predictions could be more accurate as averaging the price over 15 neighbors can be too large of a generalization)
- Experiment with variable creations (actually test with K is best for the CompPrice KNN and evaluate model performance directly after variable creation)